cli/mender-artifact/dump.go

```diff
@@ -0,0 +1,242 @@
+	args []string
+}
+
+type dumpFileStore struct {
+	fileDir string
+	args    *[]string
+}
+
+func DumpCommand(c *cli.Context) error {
+	var dump artifactDump
+
+	if len(c.Args()) != 1 {
+		return cli.NewExitError("Need to specify exactly one Artifact with dump command",
+			errArtifactInvalidParameters)
+	}
+
+	art, err := os.Open(c.Args()[0])
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf(
+			"Error opening Artifact: %s", err.Error()),
+			errArtifactOpen)
+	}
+	defer art.Close()
+
+	ar := areader.NewReader(art)
+
+	scriptsReadCallback := func(r io.Reader, i os.FileInfo) error {
+		fullPath := path.Join(c.String("scripts"), i.Name())
+		script, err := os.OpenFile(fullPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
+		if err != nil {
+			return err
+		}
+		defer script.Close()
+
+		_, err = io.Copy(script, r)
+		if err != nil {
+			return err
+		}
+
+		dump.args = append(dump.args, "--script", fullPath)
+
+		return nil
+	}
+	if len(c.String("scripts")) > 0 {
+		err = os.MkdirAll(c.String("scripts"), 0755)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf(
+				"Could not create directory: %s", err.Error()), errSystemError)
+		}
+		ar.ScriptsReadCallback = scriptsReadCallback
+	}
+
+	err = ar.ReadArtifactHeaders()
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error dumping Artifact: %s",
+			err.Error()), errArtifactInvalid)
+	}
+
+	err = dumpPayloads(c, ar, &dump)
+	if err != nil {
+		return err
+	}
+
+	if c.Bool("print-cmdline") {
+		printCmdline(ar, dump.args)
+	}
+
+	return nil
+}
+
+func dumpPayloads(c *cli.Context, ar *areader.Reader, dump *artifactDump) error {
+	handlers := ar.GetHandlers()
+	if len(handlers) != 1 {
+		return cli.NewExitError("Only exactly one payload is supported for dump command",
+			errArtifactUnsupportedFeature)
+	}
+
+	if len(c.String("meta-data")) > 0 {
+		err := os.MkdirAll(c.String("meta-data"), 0755)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf(
+				"Unable to create directory: %s", err.Error()), errSystemError)
+		}
+
+		// Hardcode to 0 index for now.
+		handler := handlers[0]
+
+		for _, augmented := range []bool{false, true} {
+			var metaData map[string]interface{}
+			var fullPath string
+			var metaDataArg string
+			if augmented {
+				metaData = handler.GetUpdateAugmentMetaData()
+				fullPath = path.Join(c.String("meta-data"), "0000.meta-data-augment")
+				metaDataArg = "--augment-meta-data"
+			} else {
+				metaData = handler.GetUpdateOriginalMetaData()
+				fullPath = path.Join(c.String("meta-data"), "0000.meta-data")
+				metaDataArg = "--meta-data"
+			}
+
+			if len(metaData) == 0 {
+				continue
+			}
+
+			metaDataFd, err := os.OpenFile(fullPath,
+				os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+			if err != nil {
+				return cli.NewExitError(fmt.Sprintf(
+					"Unable to create meta-data file: %s", err.Error()), errSystemError)
+			}
+			defer metaDataFd.Close()
+
+			w := json.NewEncoder(metaDataFd)
```

Comment:
Also here I'd suggest using `json.Marshal`, but it is not so verbose here. So your free to choose.

Reply:
Again, we are writing to a stream, so we cannot use `Marshal`, which only accepts `[]byte`.